agent/taskresource/asmsecret/asmsecret.go

```diff
@@ -0,0 +1,426 @@
+func (secret *ASMSecretResource) updateAppliedStatusUnsafe(knownStatus resourcestatus.ResourceStatus) {
+	if secret.appliedStatus == resourcestatus.ResourceStatus(ASMSecretStatusNone) {
+		return
+	}
+
+	// Check if the resource transition has already finished
+	if secret.appliedStatus <= knownStatus {
+		secret.appliedStatus = resourcestatus.ResourceStatus(ASMSecretStatusNone)
+	}
+}
+
+// SetAppliedStatus sets the applied status of resource and returns whether
+// the resource is already in a transition
+func (secret *ASMSecretResource) SetAppliedStatus(status resourcestatus.ResourceStatus) bool {
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	if secret.appliedStatus != resourcestatus.ResourceStatus(ASMSecretStatusNone) {
+		// return false to indicate the set operation failed
+		return false
+	}
+
+	secret.appliedStatus = status
+	return true
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (secret *ASMSecretResource) GetKnownStatus() resourcestatus.ResourceStatus {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.knownStatusUnsafe
+}
+
+// StatusString returns the string of the cgroup resource status
+func (secret *ASMSecretResource) StatusString(status resourcestatus.ResourceStatus) string {
+	return ASMSecretStatus(status).String()
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (secret *ASMSecretResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	secret.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (secret *ASMSecretResource) GetCreatedAt() time.Time {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.createdAt
+}
+
+// It spins up multiple goroutines in order to retrieve values in parallel.
+func (secret *ASMSecretResource) Create() error {
+
+	// To fail fast, check execution role first
+	executionCredentials, ok := secret.credentialsManager.GetTaskCredentials(secret.getExecutionCredentialsID())
+	if !ok {
+		// No need to log here. managedTask.applyResourceState already does that
+		err := errors.New("asm secret resource: unable to find execution role credentials")
+		secret.setTerminalReason(err.Error())
+		return err
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	var wg sync.WaitGroup
+
+	// Get the maximum number of errors be returned, which will be one error per goroutine
+	errorEvents := make(chan error, len(secret.requiredSecrets))
+
+	seelog.Infof("asm secret resource: retrieving secrets for containers in task: [%s]", secret.taskARN)
+	secret.secretData = make(map[string]string)
+
+	for _, asmsecret := range secret.getRequiredSecrets() {
+		wg.Add(1)
+		// Spin up goroutine per secret to speed up processing time
+		go secret.retrieveASMSecretValue(asmsecret, iamCredentials, &wg, errorEvents)
```

Comment:
so what's the upperbound on the number of concurrent service calls we'd be making here?

Reply:
I have been thinking about this problem when I was implementing. So what's the main concern about not setting upper limit for it? If from the memory consumed, goroutine itself will not really consume much of the memory. If from hitting the throttling limit of AWS secrets manager API call,  it may be an issue, since it does have the limit of 700 per account per second. But customers should be able to pay and increate the limit if they need.

Reply:
main concern for me is just leaving room for unbounded growth for the number of service calls. im not completely sure what a reasonable number here would be tbh. 

i forget what the api limits are around secrets/container/task. but it useful to have a clear idea of the number of concurrent run tasks it would take to hit the aws secrets manager api limit.

Reply:
should we test this with an arbitrarily high number of secrets?